### DISCUSSION TITLE
Concise Class

### DISCUSSION THREAD

**Discussion by D2O84** [dotnet/csharplang#5330] [Language Ideas]
Similar to records but with class
`class Person(string Name, DateTime BirthDate);`

Would be equivalent to

```
class Person
{
    public Person(string name, DateTime birthDate)
    {
        Name = name;
        BirthDate = birthDate;
    }

    public string Name { get; set; }

    public DateTime BirthDate{ get; set; }
}
```

And ideally with a code refactor to convert the concise version to the full version

Would be useful for scaffolding and for simple classes

Thank you, 
Continue the great work

**Comment by HaloFour** [comment]
Primary constructors for non-records are a separate championed proposal but they would not generate public mutable data members.  This sounds like something you could do with a source generator.

**Reply by D2O84** [reply]
when you say source generator are you talking about t4?, because the idea would be to keep the concise version, and ideally out of the box without any addin's

**Reply by theunrepentantgeek** [reply]
No.

Source Generators integrate into the Roslyn compiler, can be distributed via NuGet packages and more.

Here's a starting point: [https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/)

**Reply by D2O84** [reply]
Not too sure that apply, the actual line of code and all references to that class would be marked as error from visual studio, before compilation
`class Person(string Name, DateTime BirthDate);`

**Reply by theunrepentantgeek** [reply]
A source generator won't enable new syntax - but it can solve the problem, and it can do so today.

Off the top of my head, you could write something like this:

``` csharp
// Attribute to trigger the source generator
[AutoClass] 
public class Person
{
    public string Name { get; set; }
    public DateOnly BirthDate { get; set; }
}
```

The source generator would then inject the constructor.

**Reply by KarolKalbarczyk** [reply]
Person would have to partial in order for source generator to be able to add a constructor.

**Reply by D2O84** [reply]
but the point is not to generate the constructor, it's to declare a class concisely, like this
`class Person(string Name, DateTime BirthDate);`
no `{ get; set; }`, no property scope, no extra curly braces
I don't need this feature today, but this feature would reduce my code size while improving clarity and readability.

**Reply by HaloFour** [reply]
Correct, you couldn't achieve the exact syntax you are proposing, but you can get relatively close.  This would work with record primary constructors today by having the source generator re-emit the property members:

```cs
// you write something like this
[AutoClass]
public partial record Person(string Name, DateOnly BirthDate);

// source generator emits something like this
public partial record Person
{
    public String Name { get; set; } = Name;
    public DateTime BirthDate { get; set; } = BirthDate;
}
```

Here's the current proposal for non-record primary constructors: https://github.com/dotnet/csharplang/blob/main/proposals/non-record-primary-constructors.md

**Reply by D2O84** [reply]
From what I understood of source generator, you cannot delete/modify code so I would end up with 2 record/class definition (one record one class). 
Also the current proposal seems very different from the record version, if I understand well the class parameters doesn't generate corresponding Properties, it just a primary constructor with parameters and auto generated fields.
Why do it differently than records, which is simpler, clearer and it would be consistent with records, if I create a record and then I want to change it to a class I just have to change the keyword record to class and vice versa?

**Reply by HaloFour** [reply]
> From what I understood of source generator, you cannot delete/modify code so I would end up with 2 record/class definition (one record one class).

Records permit the members to be explicitly specified in which case the compiler will not automatically emit that member.  If you provide a mutable version of the property then the compiler will not generate an init-only version of the property.

[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKAIHoiACATwHsBXUgdwCcoAXAU1IGdKxXmALWAHNSAGygBrdvygcC6AMykADhAbMoEEaQasAxpQYJSABVYMuMABTpsABlIA5CDzikAIhDYAVKD1IAhKDU+DzYASgBuAmIyLmoGXXZBVhgzTwNSVjAWDk5uXgEYYTFJUmlZfAVlVXVNbT0DI1NzShgCAG8CUm7SKoBlZiYix2d2dtJk5gjOXmmAX1IAXhGeKPwe3sVQ1h8/QODt0nHJ6Y5Z0gXl/f5ttbnowkrsADZezF7sAHYjrp6bV/QABZSABZaBWMI/dYbHoAN1UyjMFiWpFSdBMSNalgARE4eNjXNtdqwAHReSgICDkSK/GGkEikHhgYBmUgyUgQXTMaiaETkRnUZgQYAiVi0mFKTEwEl49jLbEAeX4rNl2LWG3u+DmQA=)

> Why do it differently than records

Because they are different from records.  The `record` modifier is what indicates that the class should behave as a data carrier with value identity.  Without that modifier there's nothing to indicate that the primary constructor parameters should be exposed.  This is true of how primary constructors work in numerous other languages as well, such as Scala or Kotlin.

**Reply by 333fred** [reply]
> Why do it differently than records

The LDM meet [last week](https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-10-20.md#primary-constructors) to talk about primary constructors. This is what the notes said:

> Unlike for records, primary constructors aren't about defining the members that make up a grouping of data: instead, they're purely for defining the inputs to a class. So things like deconstruction, equality, and public properties aren't automatically generated from the parameters. Instead, they become fields, that are then eliminted if the field is never used outside of a field initializer.

**Reply by D2O84** [reply]
Regarding

> For each parameter in the parameter_list, if the type declaration does not directly contain a property or field declaration of the same name as the parameter, and if any expression within the body of the type declaration would reference such a member, then a private field is implicitly declared with the same name and type as the parameter.

would it be possible to have a public scope before the parameter that would implicitly generate a public property for that parameter

`class Person(public string Name, public DateTime BirthDate);`

**Reply by CyrusNajmabadi** [reply]
@D2O84 this week be discussed at an upcoming meeting.

**Comment by DavidArno** [comment]
Given the use-case described here, it's clear that different developers will have different requirements for primary constructors. For example, given the following code:
```c#
interface IPerson
{
    string Name { get; set; }
    DateTime BirthDate{ get; }
}

class Person(string Name, DateTime BirthDate) : IPerson;
```
It seems reasonable to expect the lowered code to be:
```c#
class Person : IPerson
{
    public Person(string Name, DateTime BirthDate)
    {
        this.Name = Name;
        this.BirthDate = BirthDate;
    }

    public string Name { get; set; }

    public DateTime BirthDate{ get; }
}
```
Having gone around in circles a bit on this one in my own head, I've definitely settled on the idea that that primary constructor is just a shorthand notation for the constructor. With no prompting, it should go for the "pit of success" option of read-only private fields mapping to the constructor parameter. Different scenarios can then be supported by the developer specifying their own field/property requirements. That could be done via an interface as per the above example, or in the OP's case by specifying:
```c#
class Person(string Name, DateTime BirthDate)
{
    public string Name { get; set; }

    public DateTime BirthDate{ get; set; }
}
```
